lib/ansible/cli/doc.py

```diff
@@ -604,44 +604,44 @@ def _get_collection_filter(self):
     def _list_plugins(self, plugin_type, content):
 
         results = {}
+        self.plugins = {}
         loader = DocCLI._prep_loader(plugin_type)
 
         coll_filter = self._get_collection_filter()
-        if coll_filter in ('ansible.builtin', 'ansible.legacy', '', None):
-            paths = loader._get_paths_with_context()
-            for path_context in paths:
-                self.plugin_list.update(DocCLI.find_plugins(path_context.path, path_context.internal, plugin_type))
-
-        add_collection_plugins(self.plugin_list, plugin_type, coll_filter=coll_filter)
+        self.plugins.update(list_plugins(plugin_type, coll_filter))
 
         # get appropriate content depending on option
         if content == 'dir':
             results = self._get_plugin_list_descriptions(loader)
         elif content == 'files':
-            results = self._get_plugin_list_filenames(loader)
+            results = {k: self.plugins[k][0] for k in self.plugins.keys()}
         else:
-            results = {k: {} for k in self.plugin_list}
+            results = {k: {} for k in self.plugins.keys()}
             self.plugin_list = set()  # reset for next iteration
+
         return results
 
-    def _get_plugins_docs(self, plugin_type, names):
+    def _get_plugins_docs(self, plugin_type, names, fail_ok=False):
```

Comment:
The default `fail_ok=False` will break the ansible-doc sanity test (which is really important for collections, since it's basically the only documentation test for plugins in collections right now).